dapp/src/utils/hooks.js

```diff
@@ -1,57 +1,103 @@
+  getConnector,
+  getConnectorImage,
+} from './connectors'
 import AccountStore from 'stores/AccountStore'
 import analytics from 'utils/analytics'
 
 export function useEagerConnect() {
   const { activate, active } = useWeb3React()
 
-  const [tried, setTried] = useState(false)
+  const [triedInjected, setTriedInjected] = useState(false)
+  const [triedSafeMultisig, setTriedSafeMultisig] = useState(false)
+  const [safeMultisigConnector, setSafeMultisigConnector] = useState(null)
+  const [connector, setConnector] = useState(null)
+  const [connectorIcon, setConnectorIcon] = useState(null)
+
+  // Instantiate Gnosis Safe web3-react connector
+  useEffect(() => {
+    // Requires global.window to be available to must be instantiated inside
+    // useEffect for Next.js to work
+    setSafeMultisigConnector(new SafeAppConnector())
+  }, []) // Do this on mount
+
+  // Attempt to use Gnosis Safe Multisig if available
+  useEffect(() => {
+    async function attemptSafeConnection() {
+      // Safe multisig connector not yet initialised, wait for it
+      if (!safeMultisigConnector) return
+
+      // OK to use Gnosis Safe?
+      const canUseGnosisSafe =
+        safeMultisigConnector && (await safeMultisigConnector.isSafeApp())
+
+      try {
+        await activate(safeMultisigConnector, undefined, true)
+      } catch (error) {
+        // Outside of Safe context
+        console.debug(error)
+        setTriedSafeMultisig(true)
+        return
+      }
+
+      setConnector(safeMultisigConnector)
+      setConnectorIcon('gnosis-safe-icon.svg')
+
+      setTriedSafeMultisig(true)
+    }
+
+    attemptSafeConnection()
+  }, [safeMultisigConnector]) // Try this when Safe multisig connector is started
 
+  // Attempt to use injected connector
   useEffect(() => {
-    if (tried || localStorage.getItem('eagerConnect') === 'false') return
-
-    //TODO: solve for other connectors
-    injected.isAuthorized().then((isAuthorized) => {
-      if (isAuthorized) {
-        activate(injected, undefined, true)
-          .then(() => {
-            const connectorName = Object.keys(connectorsByName).filter(
-              (cKey) => {
-                return connectorsByName[cKey].connector === injected
-              }
-            )[0]
-            analytics.track('Wallet connected', {
-              vendor: connectorName,
-              eagerConnect: true,
-            })
-
-            AccountStore.update((s) => {
-              s.connectorIcon = getConnectorImage(
-                connectorsByName[connectorName]
-              )
-            })
-          })
-          .catch((e) => {
-            console.error(e)
-            setTried(true)
-          })
-      } else {
-        setTried(true)
+    async function attemptInjectedConnection() {
+      // Must try Safe multisig before injected connector
+      if (!triedSafeMultisig) return
+      // Already got a connector, don't try another, use icon as indicator
+      if (connectorIcon) return
```

Comment:
IMHO would be better to have a separate bool indicator for this. Since it is not obvious that changing `connectorIcon ` logic prevents this function from trying a non gnosis login.

Reply:
Done